multipaz-compose/src/commonMain/kotlin/org/multipaz/compose/document/DocumentModel.kt

```diff
@@ -0,0 +1,144 @@
+
+                }
+                .launchIn(scope)
+        }
+    }
+
+    private suspend fun updateDocumentInfo(
+        documentId: String? = null,
+        event: DocumentEvent? = null,
+    ) {
+        val id = event?.documentId ?: documentId ?: return
+        if (event is DocumentDeleted) {
+            _documentInfos.update { current ->
+                LinkedHashMap(current).apply {
+                    remove(id)
+                }
+            }
+        } else {
+            documentStore.lookupDocument(id)?.let { document ->
+                _documentInfos.update { current ->
+                    val newMap = LinkedHashMap<String, DocumentInfo>(current.size + 1)
+
+                    current.forEach { (existingDocumentId, existingDocumentInfo) ->
+                        newMap[existingDocumentId] =
+                            if (existingDocumentId == id) document.toDocumentInfo()
+                            else existingDocumentInfo
+                    }
+
+                    if (!current.containsKey(id)) {
+                        newMap[id] = document.toDocumentInfo()
+                    }
+
+                    newMap
+                }
+            }
+        }
+    }
+
+
+    companion object {
+        private const val TAG = "DocumentModel"
+
+        private suspend fun Document.buildCredentialInfos(): List<CredentialInfo> {
+            return getCredentials().map { credential ->
+                val keyInfo = if (credential is SecureAreaBoundCredential) {
+                    credential.secureArea.getKeyInfo(credential.alias)
+                } else {
+                    null
+                }
+                val keyInvalidated = if (credential is SecureAreaBoundCredential) {
+                    credential.secureArea.getKeyInvalidated(credential.alias)
+                } else {
+                    false
+                }
+                CredentialInfo(
+                    credential = credential,
+                    keyInfo = keyInfo,
+                    keyInvalidated = keyInvalidated
+                )
+            }
+        }
+
+        private suspend fun Document.toDocumentInfo(): DocumentInfo {
+            val cardArtBytes = metadata.cardArt?.toByteArray()
+                ?: Res.drawable.default_card_art.toByteArray()
+            Logger.d("name", "${metadata.displayName}")
+            Logger.d("typename", "${metadata.typeDisplayName}")
+            val updatedCardArtBytes = if (metadata.cardArt == null) {
+                overLayWithText(
+                    cardArtBytes,
+                    metadata.displayName,
+                    metadata.typeDisplayName
+                )
+
+            } else {
+                cardArtBytes
+            }
+            return DocumentInfo(
+                document = this,
+                cardArt = decodeImage(updatedCardArtBytes),
+                credentialInfos = buildCredentialInfos()
+            )
```

Comment:
As mentioned in the other comment, move all the text-overlay so it lives in this file. This should also avoid an extra image encoding/decoding step for when there's no cardArt.